taichi/rhi/metal/metal_device.mm

```diff
@@ -0,0 +1,495 @@
+    [mtl_library_ release];
+    is_destroyed_ = true;
+  }
+}
+
+MetalShaderResourceSet::MetalShaderResourceSet(const MetalDevice &device)
+    : device_(&device) {}
+MetalShaderResourceSet::~MetalShaderResourceSet() {}
+
+ShaderResourceSet &MetalShaderResourceSet::buffer(uint32_t binding,
+                                                  DevicePtr ptr, size_t size) {
+  TI_ASSERT(ptr.device == (Device *)device_);
+  const MetalMemory &memory = device_->get_memory(ptr.alloc_id);
+
+  MetalShaderResource rsc{};
+  rsc.ty = MetalShaderResourceType::buffer;
+  rsc.binding = binding;
+  rsc.buffer.buffer = memory.mtl_buffer();
+  rsc.buffer.offset = ptr.offset;
+  rsc.buffer.size = size;
+  resources_.emplace_back(std::move(rsc));
+
+  return *this;
+}
+ShaderResourceSet &MetalShaderResourceSet::buffer(uint32_t binding,
+                                                  DeviceAllocation alloc) {
+  TI_ASSERT(alloc.device == (Device *)device_);
+  const MetalMemory &memory = device_->get_memory(alloc.alloc_id);
+
+  MetalShaderResource rsc{};
+  rsc.ty = MetalShaderResourceType::buffer;
+  rsc.binding = binding;
+  rsc.buffer.buffer = memory.mtl_buffer();
+  rsc.buffer.offset = 0;
+  rsc.buffer.size = memory.size();
+  resources_.emplace_back(std::move(rsc));
+
+  return *this;
+}
+
+ShaderResourceSet &MetalShaderResourceSet::rw_buffer(uint32_t binding,
+                                                     DevicePtr ptr,
+                                                     size_t size) {
+  TI_ASSERT(ptr.device == (Device *)device_);
+  const MetalMemory &memory = device_->get_memory(ptr.alloc_id);
+
+  MetalShaderResource rsc{};
+  rsc.ty = MetalShaderResourceType::buffer;
+  rsc.binding = binding;
+  rsc.buffer.buffer = memory.mtl_buffer();
+  rsc.buffer.offset = ptr.offset;
+  rsc.buffer.size = size;
+  resources_.emplace_back(std::move(rsc));
+
+  return *this;
+}
+ShaderResourceSet &MetalShaderResourceSet::rw_buffer(uint32_t binding,
+                                                     DeviceAllocation alloc) {
+  TI_ASSERT(alloc.device == (Device *)device_);
+  const MetalMemory &memory = device_->get_memory(alloc.alloc_id);
+
+  MetalShaderResource rsc{};
+  rsc.ty = MetalShaderResourceType::buffer;
+  rsc.binding = binding;
+  rsc.buffer.buffer = memory.mtl_buffer();
+  rsc.buffer.offset = 0;
+  rsc.buffer.size = memory.size();
+  resources_.emplace_back(std::move(rsc));
+
+  return *this;
+}
+
+MetalCommandList::MetalCommandList(const MetalDevice &device)
+    : device_(&device) {}
+MetalCommandList::~MetalCommandList() {}
+
+void MetalCommandList::bind_pipeline(Pipeline *p) noexcept {
+  TI_ASSERT(p != nullptr);
+  current_pipeline_ = (MetalPipeline *)p;
+}
+RhiResult MetalCommandList::bind_shader_resources(ShaderResourceSet *res,
+                                                  int set_index) noexcept {
+  TI_ASSERT(res != nullptr);
+  TI_ASSERT(set_index == 0);
+  current_shader_resource_set_ = (MetalShaderResourceSet *)res;
```

Comment:
This command always happen after bind pipeline because bind pipeline will flush all bound resources

Reply:
So I don't think it's the best idea to do current_shader_resource_set_